lib/floe/workflow/states/map.rb

```diff
@@ -77,27 +85,46 @@ def end?
         end
 
         def ready?(context)
-          !context.state_started? || context.state["ItemProcessorContext"].any? { |ctx| item_processor.step_nonblock_ready?(Context.new(ctx)) }
+          !context.state_started? || each_item_processor(context).any? { |ctx| item_processor.step_nonblock_ready?(ctx) }
         end
 
         def wait_until(context)
-          context.state["ItemProcessorContext"].filter_map { |ctx| item_processor.wait_until(Context.new(ctx)) }.min
+          each_item_processor(context).filter_map { |ctx| item_processor.wait_until(ctx) }.min
         end
 
         def waiting?(context)
-          context.state["ItemProcessorContext"].any? { |ctx| item_processor.waiting?(Context.new(ctx)) }
+          each_item_processor(context).any? { |ctx| item_processor.waiting?(ctx) }
         end
 
         def running?(context)
           !ended?(context)
         end
 
         def ended?(context)
-          context.state["ItemProcessorContext"].all? { |ctx| Context.new(ctx).ended? }
+          each_item_processor(context).all?(&:ended?)
+        end
+
+        def success?(context)
+          contexts   = each_item_processor(context)
+          num_failed = contexts.count(&:failed?)
+          total      = contexts.count
+
+          return true if num_failed.zero?
+
+          # Some have failed, check the tolerated_failure thresholds to see if
+          # we should fail the whole state.
+          return true if tolerated_failure_count      && num_failed < tolerated_failure_count
+          return true if tolerated_failure_percentage && (100 * num_failed / total.to_f) < tolerated_failure_percentage
```

Comment:
What is supposed to happen if both count and percentage are given (or is that not allowed)?  I can think of percentages where you want, say, the minimum of 50% or 3.  In that case, I think you need to check both clauses, and then return, as opposed to bailing out on the first one.

(This could have fallen through with the flip from failed? to success?)

Reply:
specs tend to say one vs the other.
In the branch that I had with all error checkings, I only allow or the other.

Also, you should not be able to state Next and End at the same time.
But in the short term, we've been letting these cases slide

Reply:
From the states language spec:

> If a "ToleratedFailurePercentage" field and a "ToleratedFailureCount" field are both specified, the Map State will fail if either threshold is breached.

Reply:
There's an additional nuance here that I'm wondering if we need to code explicitly (not 100% sure).  The spec says

> A Map State MAY have a "ToleratedFailurePercentage" field whose value MUST be a number between zero and 100. Its default value is zero, which means the Map State will fail if any (i.e. more than 0%) of its items fail. A "ToleratedFailurePercentage" value of 100 means the interpreter will continue starting iterations even if all items fail.

So if 0 or 100 are specified, then those have defined meanings, and I'm concerned that silly floating point math might let those fall through the cracks?  I wonder if we should have at least some tests for those specific values.

Reply:
Another very strange question but is it possible for `total` to be 0 here or does some earlier check avoid that?  Asking because this is a potential divide by zero here.

Reply:
> From the states language spec:
>> If a "ToleratedFailurePercentage" field and a "ToleratedFailureCount" field are both specified, the Map State will fail if either threshold is breached.

Right this will report a failure if either threshold is hit

Reply:
> Another very strange question but is it possible for total to be 0 here or does some earlier check avoid that? Asking because this is a potential divide by zero here.

Technically this will return early if total is zero because num_failed will also be zero, but I can add an additional / explicit total.zero? above

Reply:
> Right this will report a failure if either threshold is hit

The way it's coded though, I'm not sure it does?  Taking a (very) contrived example, if we had 4 items, 2 failures,  threshold count of 2, and the threshold % of 25%, then the current code will return success true, but should return false.

Reply:
Okay I added an explicit check for ToleratedFailurePercentage==100 (interesting the spec says it is an integer not a float so that made `== 100` easier, I did `&.to_i` in the initialize)
I think all concerns here are convered.

Reply:
> The way it's coded though, I'm not sure it does? Taking a (very) contrived example, if we had 4 items, 2 failures, threshold count of 2, and the threshold % of 25%, then the current code will return success true, but should return false.

Oh yeah, this did flip when going from `failed?` to `success?`